### PR TITLE
[POC] Add caching for faketensor propagation

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109485
* #113303

### CODE REVIEW

**Comment by Chillee** [comment]
From my experimental testing, seems like this can reduce E2E compilation time by about 20%, and can reduce tracing time by a significant amount as well.

**Comment by ezyang** [comment]
cc @ysiraichi

**Review by eellison — COMMENTED** [review]
i know this is a wip, but leaving some comments

**Comment by Chillee** [comment]
Seems like this reduces compilation time on our dashboard by about 15% on average on timm and about 20% on HF.

**Comment by Chillee** [comment]
1. Make the hashing/caching actually work - need to cache on a lot more properties. Look at dynamo hashing of tensor properties - there might be something we can reuse.
2. Handle outputs that return more than one tensor.
3. Handle aliasing of inputs and outputs appropriately.
4. Optimize the hashing part.

Long term plan:
1. Optimize symbolic shape propagation.

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -1226,6 +1226,52 @@ def merge_devices(t):
...
+                    output_tensor.stride(),
+                    output_tensor.dtype,
+                    output_tensor.device,
+                )
+                cache[arg_hash] = metadata
+            return output_tensor
+        else:
+            metadata = cache[arg_hash]
+
+            return FakeTensor(
```
[the remark this answers is not in the document]
Comment (ysiraichi) [reply torch/_subclasses/fake_tensor.py:1261]: @eellison Could you clarify why we should do that? As far as I understand:

[the remark this answers is not in the document]
Comment (eellison) [reply torch/_subclasses/fake_tensor.py:1261]: constants: https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/fake_tensor.py#L1062 fake tensor has a constant field to keep track of things 1-elem tensors so we can propagate through `x[:torch.tensor(5)]` without dynamic shapes.

non-symbolic tensors: yes, what you said, and additionally, we would need to make sure the right guards get added

[the remark this answers is not in the document]
Comment (ysiraichi) [reply torch/_subclasses/fake_tensor.py:1261]: So, we could include fake tensors that have constants if we include it in the hash key, right!?

**Pull request closed without merging**